cupy/linalg/decomposition.py

```diff
@@ -170,7 +170,102 @@ def qr(a, mode='reduced'):
     return q, _triu(r)
 
 
-# TODO(okuta): Implement svd
+def svd(a, full_matrices=True, compute_uv=True):
+    '''Singular Value Decomposition.
+
+    Factorizes the matrix ``a`` as ``u * np.diag(s) * v``, where ``u`` and
+    ``v`` are unitary and ``s`` is an one-dimensional array of ``a``'s
+    singular values.
+
+    Args:
+        a (cupy.ndarray): The input matrix with dimension ``(M, N)``.
+        full_matrices (bool): If True, it returns U and V with dimensions
+            ``(M, M)`` and ``(N, N)``. Otherwise, the dimensions of U and V
+            are respectively ``(M, K)`` and ``(K, N)``, where
+            ``K = min(M, N)``.
+        compute_uv (bool): If True, it only returns singular values.
+
+    .. seealso:: :func:`numpy.linalg.svd`
+    '''
+    if not cuda.cusolver_enabled:
+        raise RuntimeError('Current cupy only supports cusolver in CUDA 8.0')
+
+    # TODO(Saito): Current implementation only accepts two-dimensional arrays
+    _assert_cupy_array(a)
+    _assert_rank2(a)
+
+    # Cast to float32 or float64
+    if a.dtype.char == 'f' or a.dtype.char == 'd':
+        dtype = a.dtype.char
+    else:
+        dtype = numpy.find_common_type((a.dtype.char, 'f'), ()).char
+
+    # Remark 1: gesvd only supports m >= n (WHAT?)
+    # Remark 2: gesvd only supports jobu = 'A' and jobvt = 'A'
+    # Remark 3: gesvd returns matrix U and V^H
+    # Remark 4: Remark 2 is removed since cuda 8.0 (new!)
+    n, m = a.shape
+    if m >= n:
+        x = a.astype(dtype, copy=True)
+        trans_flag = False
+    else:
+        m, n = a.shape
+        x = a.transpose().astype(dtype, copy=True)
+        trans_flag = True
+    mn = min(m, n)
+
+    if compute_uv:
+        if full_matrices:
+            u = cupy.empty((m, m), dtype=dtype)
+            vt = cupy.empty((n, n), dtype=dtype)
+        else:
+            u = cupy.empty((mn, m), dtype=dtype)
+            vt = cupy.empty((mn, n), dtype=dtype)
+        u_ptr, vt_ptr = u.data.ptr, vt.data.ptr
+    else:
+        u_ptr, vt_ptr = 0, 0  # Use nullptr
+    s = cupy.empty(mn, dtype=dtype)
+    handle = device.get_cusolver_handle()
+    devInfo = cupy.empty(1, dtype=numpy.int32)
+    if compute_uv:
+        if full_matrices:
+            jobu, jobvt = ord('A'), ord('A')
```

Comment:
We do not need to use two variables in this case, but I just wanted to clarify that `jobu` and `jobvt` are always the same.
```python
    if compute_uv:
        job = ord('A') if full_matrices else ord('S')
    else:
        job = ord('N')
```
is also OK.